packages/financial-templates-lib/src/price-feed/InsuredBridgePriceFeed.ts

```diff
@@ -0,0 +1,194 @@
+import assert from "assert";
+import { PriceFeedInterface } from "./PriceFeedInterface";
+import Web3 from "web3";
+import { parseAncillaryData } from "@uma/common";
+import { getAbi } from "@uma/contracts-node";
+import { BN } from "../types";
+import type { Logger } from "winston";
+import { InsuredBridgeL1Client } from "../clients/InsuredBridgeL1Client";
+import { InsuredBridgeL2Client } from "../clients/InsuredBridgeL2Client";
+
+const { toBN, isAddress } = Web3.utils;
+
+enum relayState {
+  Pending,
+  SpedUp,
+  Disputed,
+  Finalized,
+}
+
+enum isRelayValid {
+  Yes,
+  No,
+}
+
+interface Params {
+  logger: Logger;
+  web3: Web3;
+  bridgeAdminAddress: string;
+}
+
+interface Relay {
+  bridgePoolAddress: string;
+  quoteTimestamp: number;
+  depositId: number;
+  sender: string;
+  slowRelayer: string;
+  disputedSlowRelayers: string[];
+  instantRelayer: string;
+  depositTimestamp: number;
+  recipient: string;
+  l1Token: string;
+  amount: string;
+  slowRelayFeePct: string;
+  instantRelayFeePct: string;
+  realizedLpFeePct: string;
+  priceRequestAncillaryDataHash: string;
+  depositHash: string;
+  depositContract: string;
+  relayState: relayState;
+  relayTimestamp: number;
+}
+
+interface Deposit {
+  depositId: number;
+  timestamp: number;
+  sender: string;
+  recipient: string;
+  l1Token: string;
+  amount: string;
+  slowRelayFeePct: string;
+  instantRelayFeePct: string;
+  quoteTimestamp: number;
+}
+
+// Allows user to respond to a "relay" price request that was sent in response to a "deposit" on a InsuredBridge
+// deployed to an L2 network. The relay price request is submitted on L1. This pricefeed will respond True or False
+// based on whether the relay was correctly constructed to match a deposit. The price request includes parameters in
+// its ancillary data that must be parsed from the hex ancillary data.
+export class InsuredBridgePriceFeed extends PriceFeedInterface {
+  private readonly decimals: number;
+  private readonly logger: Logger;
+  private readonly web3: Web3;
+  private readonly bridgeAdminAddress: string;
+  private l1Client: InsuredBridgeL1Client | null = null;
+  private l2Client: InsuredBridgeL2Client | null = null;
+  private relays: Relay[] = [];
+  private deposits: Deposit[] = [];
+
+  /**
+   * @notice Constructs the InsuredBridgePriceFeed.
+   * @param {Object} logger Winston module used to send logs.
+   * @param {Object} web3 Provider to connect to Ethereum network.
+   * @param {String} bridgeAdminAddress Address of BridgeAdmin contract on L1.
+   * @param
+   */
+  constructor({ logger, web3, bridgeAdminAddress }: Params) {
```

Comment:
I think it would be better to feed in the `l1Client` and `l1Client` in the constructor thereby leaving the logic for creating these clients outside of the feed, within the `CreatePriceFeed.ts` utility. In doing so, the price feed does not need to know anything about how the clients are set up and it can _assume_ that it is fed a correct price feed.

If the feed is created from the disputer bot then the create price feed will make these clients. else, if the feed is created from the relayer, then these feeds are already instantiated and can be simply loaded in.

does this make sence and do you agree?

Reply:
I agree

Reply:
Changed!